atef/widgets/config/find_replace.py

```diff
@@ -0,0 +1,1009 @@
+from atef.widgets.core import DesignerDisplay
+from atef.widgets.utils import BusyCursorThread, insert_widget
+
+logger = logging.getLogger(__name__)
+
+
+@contextmanager
+def patch_client_cache():
+    old_happi_cache = happi.loader.cache
+    try:
+        happi.loader.cache = {}
+        dcache = DataCache()
+        dcache.signals.clear()
+        yield
+    finally:
+        happi.loader.cache = old_happi_cache
+        dcache.signals.clear()
+
+
+def walk_find_match(
+    item: Any,
+    match: Callable,
+    parent: List[Tuple[Any, Any]] = []
+) -> Generator:
+    """
+    Walk the dataclass and find every key / field where ``match`` evaluates to True.
+
+    Yields a list of 'paths' to the matching key / field. A path is a list of
+    (object, field) tuples that lead from the top level ``item`` to the matching
+    key / field.
+    - If the object is a dataclass, `field` will be a field in that dataclass
+    - If the object is a list, `field` will be the index in that list
+    - If the object is a dict, `field` will be a key in that dictionary
+
+    ``match`` should be a Callable taking a single argument and returning a boolean,
+    specifying whether that argument matched a search term or not.  This is
+    commonly a simple lambda wrapping an equality or regex search.
+
+    Ex:
+    paths = walk_find_match(ConfigFile, lambda x: x == 5)
+    paths = walk_find_match(ConfigFile, lambda x: re.compile('^warning$').search(x) is not None)
+
+    Parameters
+    ----------
+    item : Any
+        the item to search in.  A dataclass at the top level, but can also be a
+        list or dict
+    match : Callable
+        a function that takes a single argument and returns a boolean
+    parent : List[Tuple[Union[str, int], Any]], optional
+        the 'path' traveled to arive at ``item`` at this point, by default []
+        (used internally)
+
+    Yields
+    ------
+    List[Tuple[Any, Any]]
+        paths leading to keys or fields where ``match`` is True
+    """
+    if is_dataclass(item):
+        # get fields, recurse through fields
+        for field in fields(item):
+            yield from walk_find_match(getattr(item, field.name), match,
+                                       parent=parent + [(item, field.name)])
+    elif isinstance(item, list):
+        for idx, l_item in enumerate(item):
+            # TODO: py3.10 allows isinstance with Unions
+            if isinstance(l_item, get_args(PrimitiveType)) and match(l_item):
+                yield parent + [('__list__', idx)]
+            else:
+                yield from walk_find_match(l_item, match,
+                                           parent=parent + [('__list__', idx)])
+    elif isinstance(item, dict):
+        for d_key, d_value in item.items():
+            # don't halt at first key match, values could also have matches
+            if isinstance(d_value, get_args(PrimitiveType)) and match(d_value):
+                yield parent + [('__dictvalue__', d_key)]
```

Comment:
Checking the coverage - running test suite locally, it doesn't touch this line
(`coverage html && open htmlcov/index.html`)

GUI stuff aside, in my opinion we should try to aim for ~100% coverage on core library stuff that deals with our dataclasses. (Easier said than done of course!)

Reply:
I should do this more regularly in my code review as well as devlopment